### PR TITLE
Optimise cache responsiveness

### PR DESCRIPTION
**Pull request by mattrutherford**
Update subscriber immediately with existing data on new subscription

### CODE REVIEW

File: src/cache/mod.rs [hunk 1]
```diff
@@ -317,6 +256,91 @@ impl Cache {
...
+                .min_by(|a, b| {
+                    let y =
+                        a.1.created_at
+                            .signed_duration_since(*update_time)
+                            .num_milliseconds();
+                    let z =
+                        &b.1.created_at
+                            .signed_duration_since(*update_time)
+                            .num_milliseconds();
+                    (y * y).cmp(&(z * z))
```
Comment (dvdplm) [inline src/cache/mod.rs:321]: I'm too dumb: why square the elapsed millis, to get rid of the sign (but then why not use `abs()`?)?

File: src/cache/mod.rs [hunk 2]
```diff
@@ -317,6 +256,91 @@ impl Cache {
...
+                })
+                .expect("")
+                .0
+        }
+    };
+    let response_data = peer_message_cache.deque[idx..].to_vec();
+    let pdr = PeerDataArray {
```
Comment (dvdplm) [inline src/cache/mod.rs:328]: Does `pdr` stand for `peer_data_response`? I for one prefer longer variable names.

File: src/cache/mod.rs [hunk 3]
```diff
@@ -488,7 +513,7 @@ mod tests {
         let mut cache = get_test_setup();
         add_cache_entries(&mut cache);
         assert_eq!(cache.cache.len(), 2);
-        for (pm, pmc) in &cache.cache {
+        for (_pm, pmc) in &cache.cache {
```
Comment (dvdplm) [inline src/cache/mod.rs:516]: Does `pmc` stand for `peer_message_cache`? Future readers will thank you for clearer variable names.

File: src/db/reputation.rs [hunk 4]
```diff
@@ -289,7 +289,7 @@ impl DbExecutor {
             "QmQJmDorK9c8KjMF5PdWiH2WGUXyzJtgTeJ55S5gggdju7",
         ];
         if max < peer_ids.len() {
-            peer_ids.split_off(max);
+            let _ = peer_ids.split_off(max);
```
Comment (dvdplm) [inline src/db/reputation.rs:292]: Can you use `truncate()` here?

**Review by dvdplm — COMMENTED** [review]
Still need to try it out in real life but leaving a few comments.

**Review by insipx — APPROVED** [review]
Looks good from my end

File: src/cache/mod.rs [hunk 5]
```diff
@@ -317,6 +256,91 @@ impl Cache {
     }
 }
 
+fn send_updates(
+    peer_message: &PeerMessage,
+    peer_message_cache: &mut PeerMessageCache,
+    subscribers: &mut HashMap<Recipient<PeerDataArray>, PeerMessages>,
+) -> Vec<Recipient<PeerDataArray>> {
+    // Iterate through subscribers and send latest data based on their last_updated time
+    // Can be optimised for usual best case with fallback to binary_search
```
[the remark this answers is not in the document]
Comment (mattrutherford) [reply src/cache/mod.rs]: https://github.com/paritytech/substrate-analytics/issues/63

**Comment by mattrutherford** [comment]
Thanks for the reviews :) - I addressed all suggestions.

**Review by dvdplm — APPROVED** [review]

**Pull request merged**